Milestone: Finished Event feature

cyonii opened the pull request:
In this milestone, the following work has been completed:

- Create `User` model and it's controller actions, and implement a simple authentication system using sessions
- Create `Event` model and it's controller
- Create `Invitation` model and associate `User` and `Event`
- Design UI for app

cyonii committed:
Add Github workflows and rubocop config

cyonii committed:
Add required gems

cyonii committed:
Build User model and simple authentication system

cyonii committed:
Add Event model and it's controllers

- associate event to user
- design UI for Event views

cyonii committed:
Show part of event's description on card

cyonii committed:
Increase title font size

cyonii committed:
Add .border-3 to card

cyonii committed:
Create invitaions table

Use invitations table as a through table to associate events and 
attendees (User)

cyonii committed:
Destroy invitations when an event is destroyed

cyonii committed:
Add margin bottom to columns

cyonii committed:
Adjust container

cyonii committed:
Display list of attendees

cyonii committed:
Seed db using faker

cyonii committed:
Fix #show method

cyonii committed:
Add font-size util class

cyonii committed:
Make attendees clickable

cyonii committed:
Display a list of events user is invited

cyonii committed:
Make current_user display link to current_user show page

cyonii committed:
Update seed

cyonii committed:
Display user's past and upcoming event invitations

cyonii committed:
Display upcoming and past events accordingly

cyonii committed:
Display more details on event card

cyonii committed:
Update README

- document code

cyonii committed:
Add newline at end of file

cyonii committed:
Add utf-8 and responsive meta tags

Spayco commented on app/controllers/events_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  before_action :set_event, only: %i[show destroy edit update]
+  before_action :owner?, only: %i[destroy edit]
+
+  def index
+    @events = Event.all
+  end
```
* [ ] list past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] list ucomping events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

**_don't add `upcoming_events` and `previous_events` in the controller, these should be in the events model._**

cyonii replied:
Thank you

SpaYco reviewed (changes requested):
# Changes Requested 

Hello Team! 👋 
Awesome job so far!! 👍 you still need to make some changes though! 

## Required Changes ♻️ 

- [x]  I can't seem to find a way to attend events or add attendees, therefore I can't check some of the project specifications, please fix that.


### please make the appropriate changes and ask for another code review ✔️ 

Happy Coding!

Best, 
Aziz Mejri 
_contact me on [twitter](https://twitter.com/iSpaYco) or slack (@SpaYco) if you need any help_

cyonii committed:
Refactoring

cyonii committed:
Add attendees and cancel invite features

cyonii committed:
Fix linter error with rubocop

cyonii committed:
Render join button only for authenticated user

teekaytech commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,29 @@
+    end
+  end
+
+  def show
+    @user = User.find(params[:id])
+  end
```
- You do not necessarily have to find the user because you can easily use the `current_user` as the user in your view.
- Also, you should fetch the current user events from this action. i.e. `@created_events = current_user.events`. Therefore, you can use the `@created_events` directly in your view, instead of using `<% for event in @user.events %>` as you did in `app/views/users/show.html.erb` file.

cyonii replied:
"You do not necessarily have to find the user because you can easily use the current_user as the user in your view."

`UsersController#show` method shows the detail page of the user with the `id` in url `params` and not the detail page of the `current_user`. Kindly let me know if this is not the expected behaviour.

On the second bullet point you are saying I should make a variable in `#show` controller called `@created_events = current_user.events`, it feels like I am adding an extra unnecessary line in my code, I mean I am also going to loop through `@created_events` like `<% for event in @ created_events %>` in the `app/views/users/show.html.erb` file.

Kindly clarify on this points

teekaytech commented on app/views/layouts/application.html.erb:
```diff
@@ -5,11 +5,54 @@
+          <div class="collapse navbar-collapse" id="navbarNavLinks">
+            <div class="navbar-nav ml-auto">
+              <%= link_to 'New Event', new_event_path, class: 'nav-link border-right fw-bold' %>
+              <% if user_signed_in? %>
+                <%= link_to user_path(current_user), class: 'nav-link fw-bold' do %>
+                  Hello, <%= current_user.name %>
+                <% end %>
+                <%= link_to 'Logout', session_path(current_user.id), method: :delete, class: 'nav-link text-danger' %>
+              <% else %>
+                <%= link_to 'Login', new_session_path, class: 'nav-link' %>
+                <%= link_to 'Sign up', new_user_path, class: 'nav-link' %>
+              <% end %>
```
You should move this logic to the helper module.

cyonii replied:
Thank you.

teekaytech reviewed (changes requested):
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've done really well in this project..
However, some changes are required before it can be approved. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [x] In `users_controller` show action, you need to:
- List events previoulsy attendented by user using event model scope, i.e. `@past_events = current_user.attended_events.previous` then use `@past_events` instead of `@user.invites.where('date < ?', Date.today)` in your view.
- You should also list upcoming events previously attended by the user using event model scope, i.e. `@upcoming_events = current_user.attended_events.upcoming`, then use `@upcoming_events` instead of `@user.invites.where('date >= ?', Date.today)` in your view.

- [x] There should be absolutely NO logic in views. 
- [x] Remember about tests in your code (Unit & Integration).

- You're having some logic in your view files (i.e. `app/views/layouts/application.html.erb`, `app/views/users/_invitations.html.erb`). Kindly use the helper module to abstract view-related logic.
  - [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
  - [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
  - [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright


* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

cyonii committed:
Add default ordering to Event scope

cyonii committed:
Refactoring

cyonii committed:
Use helpers instead of logic in views

cyonii commented:
> * [ ]  Remember about tests in your code (Unit & Integration).

I have not covered any materials about how to test rails app and I really don't know how to do it. Kindly verify that I am expected to test the application.

teekaytech reviewed (changes requested):
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've done really well in making some of the changes required.
We just have to make other little changes before we move on. Let's dive in.__

**[ REQUIRED CHANGES ]**

- [x] Your users_controller#show action should have the following:
  - [x] lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  - [x] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  - [x] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [x] Adding unit and integration tests is a compulsory feature for this project. 
  - [x] models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
  - [ ] authentication have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
  - [ ] event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

- [x] You mentioned that you need me to verify that you're expected to test the application. Here's your verification from the project specification page on [pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/):

![1212](https://user-images.githubusercontent.com/26856491/103448109-fdb06a00-4c94-11eb-8006-14724d1701c9.png)
Kindly check your project specification page on pathwright. You'll find the materials needed to get started with the testing.



* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

cyonii commented:
> ## Status: Changes Required ⚠️ 🚦
> ### Hello team 👋
> **You've done really well in making some of the changes required. We just have to make other little changes before we move on. Let's dive in.**
> 
> **[ REQUIRED CHANGES ]**
> 
> * [ ]  Your users_controller#show action should have the following:
>   
>   * [ ]  lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
>   * [ ]  lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
>   * [ ]  lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
> * [ ]  Adding unit and integration tests is a compulsory feature for this project.
>   
>   * [ ]  models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
>   * [ ]  authentication have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
>   * [ ]  event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
> * [ ]  You mentioned that you need me to verify that you're expected to test the application. Here's your verification from the project specification page on [pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/):
> 
> ![1212](https://user-images.githubusercontent.com/26856491/103448109-fdb06a00-4c94-11eb-8006-14724d1701c9.png)
> Kindly check your project specification page on pathwright. You'll find the materials needed to get started with the testing.
> 
> **Please make the required changes and submit another review request when you are done.**
> 
> **Taofeek Olalere [Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
> **Happy Coding 😃**

cyonii committed:
Refactor

cyonii committed:
Refactor

- put redirect outside if-else block since it redirects to the same url 
in either case.

cyonii committed:
Add before_action for Event actions

cyonii committed:
Add integration tests for controllers

cyonii commented:
Thank you for @teekaytech for the explanation. I think I see the rationale now.

cyonii committed:
Remove unnecessary code

cyonii committed:
Update gems

- add gems for testing

cyonii committed:
Add hover style on creator link

cyonii committed:
Test models with RSpec

cyonii committed:
Test controllers with RSpec

cyonii committed:
Remove obsolete tests

PhillipUg reviewed (approved):
# Status: Project Approved! 🥇 
Congratulations on implementing all requirements and getting this project approved. :+1:

You can merge this branch into the development branch when you're ready and go on to your next project.
:bulb:  Do not forget to submit your project completion form after merging this branch.

All the best team,

For any assistance, you can contact me on slack as **Phillip Musiime**
